unibo-fc-isi-ds/lab-snippets
# [A.Y. 2024/2025 Meco, Daniel] Exercise: RPC Auth Service 1 & 2

Goal 1: Authentication

1. What’s added?
   - Users can log in with valid credentials and receive a signed token (JWT-style).
   - Tokens are validated for authenticity and expiration.

2. Why this approach?
   - Authentication is separated from other operations to ensure that every request can be traced back to an authenticated user.
   - Tokens are lightweight, stateless, and secure, making them ideal for distributed systems.

3. How does it work?
   - The server (`InMemoryAuthenticationService`) verifies credentials and generates a signed token for valid users.
   - The client includes this token in future requests to identify the user.
   - A `validate_token` method checks token validity.

---

Goal 2: Authorization

1. What’s added?
   - Certain operations (like `get_user`) are restricted to users with the admin role.
   - Unauthorized or unauthenticated attempts result in clear error messages.

2. Why this approach?
   - Authorization adds a layer of control by restricting sensitive operations to specific roles.
   - Role-based access control (RBAC) ensures that admin-only actions are protected, even in a multi-user environment.

3. How does it work?
   - For sensitive operations like `get_user`, the server checks:
     - That the provided token is valid and unexpired.
     - That the user associated with the token has the `admin` role.
   - The client includes the token in the request using the `--token` option in the CLI.

- Testing the Solution

1. Setup:
   - Start the server: `python example2_rpc_server.py 12345`.
   - Add an admin: `python example4_rpc_client_cli.py ... add -u admin -r admin -p adminpass`.
   - Authenticate as the admin and save the token.

2. Authentication Tests:
   - Authenticate: `python example4_rpc_client_cli.py ... authenticate -u admin -p adminpass`.
   - Validate token: `python example4_rpc_client_cli.py ... validate --token "<token>"`.

3. Authorization Tests:
   - Use the admin token for restricted actions:
     `python example4_rpc_client_cli.py ... get --token "<token>" -u user1`.
   - Attempt actions with no or invalid tokens and confirm the errors.

Outcome: The system securely enforces authentication and authorization, restricting sensitive operations to valid, authorized users.
